practices/appeng/adrs/0002-distributed-tracing.md

```diff
@@ -0,0 +1,156 @@
+  distributed tracing data to use at call sites (across programming
+  languages) within the codebase
+- Anti-goal: Committing to a specific "backend", i.e. platform or service for
+  gathering, exploring, and displaying trace information
+  - Sub-goal: Leaving open as many options as possible for the backend
+
+## Decision
+
+### Use OpenTelemetry
+
+- OpenTelemetry is an emerging industry standard
+  - vendors find benefit of being in the OpenTelemetry ecosystem because they
+    no longer have to create or support instrumentation libraries in an ever
+    growing array of languages, i.e. as soon as language library exists for
+    OpenTelemetry, the vendors automatically become available to support that
+    given language.
+- OpenTelemetry is vendor agnostic
+  - tracing information can be sent to hosted services (e.g. Honeycomb.io, AWS
+    X-Ray, etc) or self-hosted Open Source implementations (e.g. Zipkin, Jaeger,
+    etc)
+  - if left unconfigured, OpenTelemetry instrumentation calls default to
+    lightweight/noop executions
+- OpenTelemetry has well-maintained libraries for the languages used in the
+  layers of most Truss projects
+  - i.e. Go (back-end); JavaScript (front-end); Python (load testing); etc
+- Easily swappable back-ends
+  - e.g. could choose a local Docker version of OpenZipkin for an all-local
+    development environment
+  - e.g. can use Honeycomb.io in the experimental commercial-cloud hosted
+    environment
+  - e.g. can swap in AWS X-Ray for use in GovCloud hosted environments
+- Cons
+  - as an abstraction layer, OpenTelemetry may prohibit usage of vendor-specific
+    capabilities
+  - some OpenTelemetry libraries and tools may trail their vendor-supported
+    counterparts
+  - instrumentation for tracing may be a vector for performance overhead
+
+## Why is this Applicable to the Practice as a Whole
+
+Most applications will benefit from instrumentation once they are accepting
+production traffic. For many systems, the amount of work required to add
+tracing calls to the code is non-trivial and needs to be done preemptively.
+
+Projects should identify what types of calls need to be traced as a part of
+discovery and document these decisions in a project-level ADR. For example,
+a web backend might add tracing to HTTP requests, service invocations,
+database calls, and any network requests made by the backend.
+
+## When to Not Implement This Decision
+
+Don't use OpenTelemetry if:
+
+- the ecosystem you are working in doesn't have a mature instrumentation
+  library
+- the project already has a working instrumentation solution using a
+  proprietary hosted service that is satisfying the project's needs
+- the project is too small to support the effort required to implement
+  tracing. If this is the case, it is probably worth investigating if there
+  are other implementation options for the project, potentially in
+  higher-level languages or batteries-included frameworks.
```

Comment:
Above the ADR says a project gets the most benefit once accepting production traffic, but that adding tracing is non trivial and should be done preemptively. Is the recommendation to add it at the beginning or once a project is mature enough to be in production?

Often our projects start off as small prototypes that then grow into a larger application. Is the recommendation for a higher-level language or batteries-included framework applicable to those cases? Is there a recommendation on when to consider adding instrumentation for maximum benefit when starting with a small prototype that eventually needs to transition to a large production application?

Reply:
when I read this, I felt that _the recommendation is to add it at the beginning of a project_ is what I interpreted this as. But I think some clarification would be super helpful.

Reply:
Thanks for the feedback.  I added in some additional wording to specify that this work normally needs to occur in the early stages of the project.  I think the "small project" exclusion is really for something we don't anticipate will ever turn into something larger, not as a way to defer thinking about instrumentation until just before production.